### ISSUE TITLE
[BUG] Same limitations as Taichi for differentiability?

### ISSUE THREAD

**Issue by gpu-daal** [NVIDIA/warp#947]
### Bug Description

Taichi has an issue where kernels with global memory access (read + write) produce wrong results when calculating gradients (see https://docs.taichi-lang.org/docs/differentiable_programming#global-data-access-rules)
Same issue arises in Warp.
```
import warp as wp
wp.init()
N = 16
x = wp.array([float(i) for i in range(N)], dtype=wp.float32, requires_grad=True)
b = wp.array([10.0], dtype=wp.float32, requires_grad=True)
loss = wp.zeros(1, dtype=wp.float32, requires_grad=True)

@wp.kernel
def func_broke_rule_1(x: wp.array(dtype=wp.float32),
                      b: wp.array(dtype=wp.float32),
                      loss: wp.array(dtype=wp.float32)):
    # BAD
    loss[0] = x[1] * b[0]
    b[0] += 100.0

tape = wp.Tape()
with tape:
    wp.launch(kernel=func_broke_rule_1,
              dim=1,
              inputs=[x, b, loss])

tape.backward(loss=loss)

x_grad = x.grad.numpy()
b_grad = b.grad.numpy()
loss_val = loss.numpy()

print("loss =", loss_val[0])
print("x.grad[1] =", x_grad[1])
print("b.grad =", b_grad[0])

assert abs(x_grad[1] - 10.0) < 1e-6
```

This code is supposed to output 10.0 for x.grad[1], since for f(x) = 10*x f'(x) = 10. It outputs 110.0 instead.

### System Information

570.169, CUDA 12.8, GeForce 2080 SUPER.

**Comment by daedalus5** [comment]
@gpu-daal Yes this is a general limitation in autodiff engines. We write about it in our docs here: https://nvidia.github.io/warp/modules/differentiability.html#array-overwrite-tracking

The pattern to be avoided is WRITE after READ.

**Comment by gpu-daal** [comment]
@daedalus5 I see, do you know the root cause of this issue? I've been thinking about it for a while. Also, there is the Enzyme project (https://github.com/EnzymeAD/Enzyme) which implements AD as an IR plugin, and they don't seem to have this issue.

**Comment by daedalus5** [comment]
The primary issue is memory management. It's possible to construct a system that will automatically allocate new memory when data in the gradient chain is overwritten, but this isn't a practical approach on the GPU and Warp prefers to have users manage these situations explicitly rather than behind the scenes anyway. So in this case you can pass additional array c that stores `c[0] = b[0] + 100`

**Comment by shi-eric** [comment]
Closing this issue as it's not a bug and the question about Warp has been answered.
